File: multiprocessors/examples/01-intro-c/example8.c
// =================================================================
//
// File: example8.c
// Author(s):
// 23/08/2022
// A01705291 Ramses Aguila Flores
// A01209414 David Flores Diaz
// Description: This file contains the code that implements the
//				enumeration sort algorithm. The time this implementation
//				takes ill be used as the basis to calculate the
//				improvement obtained with parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "utils.h"

#define SIZE 50000

// implement your code

void acomodar(int *x, int *r, int s){
	for(int i=0;i<s;i++){
	int e=x[i];
	int k=0;
	for(int j=0;j<s;j++){
	if(e>x[j]){
	k++;
	}else if(e==x[j]&&i<j){
		k++;
		}
	}
	r[k]=e;
  }

}


int main(int argc, char* argv[]) {
	int i, *a, *r;
	double ms;

	r = (int*) malloc(sizeof(int) * SIZE);
	a = (int*) malloc(sizeof(int) * SIZE);
	random_array(a, SIZE);
	display_array("before", a);

	printf("Starting...\n");
	ms = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		// call the implemented function
		acomodar(a,r,SIZE);

		ms += stop_timer();
	}
	display_array("after", r);
	printf("avg time = %.5lf ms\n", (ms / N));

	free(a);
	return 0;
}

File: multiprocessors/examples/03-intro-java/Example8.java
// =================================================================
//
// File: Example8.java
// Author: Ramses Aguila Flores A01705291
//	David Flores Diaz A01209414
// Description: This file implements the enumeration sort algorithm.
// 				The time this implementation takes will be used as the
//				basis to calculate the improvement obtained with
//				parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.util.Arrays;

public class Example8 {
	private static final int SIZE = 100_000;
	private int array[];

	public Example8(int array[]) {
		this.array = array;
	}

	public void doTask() {
		int res [] = new int[SIZE];
		for (int i = 0 ; i < SIZE ; i++){
			int e = array[i];
			int k = 0;
			for (int j = 0; j < SIZE ; j++){
				if (e > array[j]) {
					k++;
				}else if( (e == array[j]) && (i < j)) {
					k++;
				}
			}
			res[k] = e;
		}
		array = res;
	}

	public int[] getSortedArray() {
		return array;
	}

	public static void main(String args[]) {
		int array[] = new int[SIZE];
		long startTime, stopTime;
		double ms;

		Utils.randomArray(array);
		Utils.displayArray("before", array);
		Example8 obj = new Example8(array);
		System.out.printf("Starting...\n");
		ms = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			obj.doTask();	// pace your code here.

			stopTime = System.currentTimeMillis();

			ms += (stopTime - startTime);
		}
		Utils.displayArray("after", obj.getSortedArray());
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
	}
}
